File: src/year2024/day1/go/part2.go

package main

import (
    "fmt"
    "strings"
    "sort"
    "strconv"
    "os"
    "bufio"
)

func main() {
    file, _ := os.Open("../prod2.txt")
    defer file.Close()

    var lines []string
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    var l, r []int

    // NOTE: see https://go.dev/blog/maps as to why need to use make() to initialize it
    counts := make(map[int]int)

    for _, line := range lines {
        nums := strings.Fields(line)
        if value, err := strconv.Atoi(nums[0]); err == nil {
            l = append(l, value)

        }

        if value, err := strconv.Atoi(nums[1]); err == nil {
            r = append(r, value)
            // This works! since an uninitialized int is 0
            counts[value]++
        }
    }

    sort.Ints(l)
    sort.Ints(r)

    var ans int = 0
    for _, left_value := range l {
        ans += (counts[left_value] * left_value)
    }

    fmt.Println("Ans: ", ans)
}

File: src/year2024/day1/scala/part2.scala

@main def part2(): Unit  =
  val lines = scala.io.Source.fromFile("../prod2.txt").getLines().toList


  var L = scala.collection.mutable.ArrayBuffer[Int]()
  var counts = scala.collection.mutable.Map[Int, Int]()
  var ans = 0

  for (line <- lines) {
    val nums = line.split(" +").map(_.toInt).toList
    L += nums(0)

    counts(nums(1)) = counts.getOrElse(nums(1), 0) + 1
  }

  for (l <- L) {
    ans += (l * counts.getOrElse(l,0))
  }

  println(ans)

